File: tests/typesafety/typescript/json_blob.ts
// expect: passes

type Json = null | boolean | number | string | Json[] | { [key: string]: Json };

export function lineTotal(line: Json): Json {
  const fields = line as { [key: string]: Json };
  return (fields.pence as number) * (fields.quantity as number);
}

export function invoiceTotal(lines: Json): Json {
  return (lines as Json[]).reduce((sum: number, row) => sum + (lineTotal(row) as number), 0);
}

export function isLarge(lines: Json): boolean {
  return (invoiceTotal(lines) as number) > 1000;
}

File: tests/typesafety/python/json_blob.py
# expect: passes
# title: Json travels through every signature, and every use needs a cast
from typing import cast

type Json = None | bool | int | float | str | list["Json"] | dict[str, "Json"]


def line_total(line: Json) -> Json:
    fields = cast(dict[str, Json], line)
    return cast(int, fields["pence"]) * cast(int, fields["quantity"])


def invoice_total(lines: Json) -> Json:
    rows = cast(list[Json], lines)
    return sum(cast(int, line_total(row)) for row in rows)


def is_large(lines: Json) -> bool:
    return cast(int, invoice_total(lines)) > 1000
